File: src/main/java/io/github/gastaldi/asciidoctor/ManagedDependencyMacro.java
package io.github.gastaldi.asciidoctor;

import org.asciidoctor.Options;
import org.asciidoctor.ast.ContentNode;
import org.asciidoctor.extension.InlineMacroProcessor;
import org.asciidoctor.extension.Name;

import java.util.Map;

@Name("version")
public class ManagedDependencyMacro extends InlineMacroProcessor {

    @Override
    public Object process(ContentNode parent, String ga, Map<String, Object> attributes) {
        Map<String, String> docAttrs = (Map<String, String>) parent.getDocument().getOptions().get(Options.ATTRIBUTES);
        String quarkusVersion = docAttrs.get("quarkus-version");
        String[] split = ga.split(":");
        String groupId = split[0];
        String artifactId = split[1];
        Resolver resolver = Resolver.forVersion(quarkusVersion);
        return resolver.findVersion(groupId, artifactId)
                .map(v -> createPhraseNode(parent, "quoted", v, attributes))
                .orElse(null);
    }
}

File: lib/quarkus/managed-dependency/managed_dependency_macro.rb
# frozen_string_literal: true
require 'rexml'
require 'open-uri'

module Quarkus::ManagedDependency
  class ManagedDependencyMacro < ::Asciidoctor::Extensions::InlineMacroProcessor
    use_dsl

    named :version
    name_positional_attributes 'quarkus-version'

    def process(parent, target, attributes)
      (group_id, artifact_id) = target.split(':')
      quarkus_version = attributes['quarkus-version'] || parent.document.attributes['quarkus-version']

      # Find what we need: http://xpather.com/85iiY0go
      dep_version_xpath_template = '//dependencyManagement/dependencies/dependency[groupId="%s" and artifactId="%s"]/version'
      quarkus_pom_url_template = 'https://repo1.maven.org/maven2/io/quarkus/quarkus-bom/%s/quarkus-bom-%s.pom'
      quarkus_pom = REXML::Document.new URI.open(quarkus_pom_url_template % [quarkus_version, quarkus_version])
      quarkus_pom.get_text dep_version_xpath_template % [group_id, artifact_id]
    end
  end
end
